src/ObjCRuntime/Registrar.cs

```diff
@@ -864,12 +1031,81 @@ public Registrar ()
 			IsDualBuild = IsDualBuildImpl;
 		}
 
-		bool IsEnum (TType type)
+		protected bool IsArray (TType type)
+		{
+			int rank;
+			return IsArray (type, out rank);
+		}
+
+		protected bool IsEnum (TType type)
 		{
 			bool dummy;
 			return IsEnum (type, out dummy);
 		}
 
+		public BindAsAttribute GetBindAsAttribute (ObjCMethod method, int parameter_index)
+		{
+			var attrib = GetBindAsAttribute (method.Method, parameter_index);
+			if (attrib != null)
+				return attrib;
+
+			if (!method.IsPropertyAccessor)
+				return null;
+
+			return GetBindAsAttribute (FindProperty (method.DeclaringType.Type, method.MethodName.Substring (4)));
+		}
+
+		bool IsSmartEnum (TType type)
+		{
+			TMethod getConstant, getValue;
+			return IsSmartEnum (type, out getConstant, out getValue);
+		}
+
+		public bool IsSmartEnum (TType type, out TMethod getConstantMethod, out TMethod getValueMethod)
+		{
+			getConstantMethod = null;
+			getValueMethod = null;
+
+			if (!IsEnum (type))
+				return false;
+
+			var extension = FindType (type, type.Namespace, type.Name + "Extensions");
+			if (extension == null)
+				return false;
+
+			var getConstantMethods = FindMethods (extension, "GetConstant");
+			foreach (var m in getConstantMethods) {
+				if (!Is (GetReturnType (m), Foundation, "NSString"))
+					continue;
+				var parameters = GetParameters (m);
+				if (parameters?.Length != 1)
+					continue;
+				if (!AreEqual (parameters [0], type))
+					continue;
+				getConstantMethod = m;
+				break;
+			}
+			if (getConstantMethod == null)
+				return false;
+
+			var getValueMethods = FindMethods (extension, "GetValue");
+			foreach (var m in getValueMethods) {
+				if (!AreEqual (GetReturnType (m), type))
+					continue;
+				var parameters = GetParameters (m);
+				if (parameters?.Length != 1)
+					continue;
+				if (!Is (parameters [0], Foundation, "NSString"))
+					continue;
+				getValueMethod = m;
+				break;
+			}
+			if (getValueMethod == null)
+				return false;
+
+			return true;
```

Comment:
This seems quite costly do to on every call. It needs numbers but I think we should cache the methods so the reflection cost only happen once per type. This can be done on a separate PR